subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java

```diff
@@ -168,17 +194,25 @@ private InstallationLocation downloadToolchain(JavaToolchainSpec spec) {
         }
     }
 
-    private JavaToolchain asToolchainOrThrow(InstallationLocation javaHome, JavaToolchainSpec spec) {
-        JavaToolchainInstantiationResult result = asToolchain(javaHome, spec, false);
-        Optional<JavaToolchain> toolchain = result.getToolchain();
-        if (!toolchain.isPresent()) {
-            JvmInstallationMetadata metadata = result.getMetadata();
+    private JavaToolchain asToolchainOrThrow(InstallationLocation javaHome, JavaToolchainSpec spec, boolean isFallback) {
+        final JvmInstallationMetadata metadata = detector.getMetadata(javaHome);
+
+        if (metadata.isValidInstallation()) {
+            return new JavaToolchain(metadata, fileFactory, new JavaToolchainInput(spec), isFallback);
+        } else {
             throw new GradleException("Toolchain installation '" + javaHome.getLocation() + "' could not be probed: " + metadata.getErrorMessage(), metadata.getErrorCause());
         }
-        return toolchain.get();
     }
 
-    private JavaToolchainInstantiationResult asToolchain(InstallationLocation javaHome, JavaToolchainSpec spec, boolean isFallback) {
-        return toolchainFactory.newInstance(javaHome, new JavaToolchainInput(spec), isFallback);
+    public class JavaToolchainInstantiationResult {
```

Comment:
Given this class can be entirely private, I would go for something shorter like `DetectedInstallation`, because the metadata is provided by the `detector`.

Reply:
:white_check_mark: